src/passes/SignaturePruning.cpp

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Remove params from signature/function types where possible.
+//
+// This differs from DeadArgumentElimination in that DAE will look at each
+// function by itself, and cannot handle indirectly-called functions. This pass
+// looks at each heap type at a time, and if all functions with a heap type do
+// not use a particular param, will remove the param.
```

Comment:
DAE disallows anything that has an indirect call. That is, it only looks at functions that are only called directly, and it considers each such function one at a time.

This pass is basically the same type of optimization but it allows indirect calls. To allow that it has to look at all the functions with the same signature, as a group (and modify them all at once, or not at all).

So the two passes are complementary, but very similar.

Reply:
Is the reason DAE cannot handle indirect calls because if a function has a possibility of being called indirectly, even though some of the arguments are not used, we cannot modify the signature, right?

So what DAE can do that this pass cannot do? I guess when two functions have the same signature and only one of them doesn't use some of the arguments, then this pass cannot do anything about that function, but DAE can remove the argument, as long as that function doesn't have a possibility of being called indirectly?

Reply:
Yes, exactly. If there are no indirect calls, then even if two functions have the same type that is not observable. So we can change one function's type while we remove a parameter in that case.